src/moonlink/src/storage/mooncake_table.rs

```diff
@@ -867,6 +881,62 @@ impl MooncakeTable {
         self.table_snapshot_watch_sender.send(lsn).unwrap();
     }
 
+    #[cfg(test)]
+    pub(crate) fn get_snapshot_watch_sender(&self) -> watch::Sender<u64> {
+        self.table_snapshot_watch_sender.clone()
+    }
+
+    /// Enter initial copy mode. Subsequent CDC events will be
+    /// buffered in a dedicated streaming memslice until
+    /// `finish_initial_copy` is called.
+    /// In this case of a streaming transaction, we simply use the already provided `xact_id` to identify the transaction. In the case of non-streaming, we use `INITIAL_COPY_XACT_ID` to identify the transaction.
+    /// All commits are buffered and deferred until initial copy finishes.
+    pub fn start_initial_copy(&mut self) {
+        self.in_initial_copy = true;
+        self.initial_copy_buffered_commits.clear();
+    }
+
+    /// Exit initial copy mode and commit all buffered changes.
+    pub async fn finish_initial_copy(&mut self) -> Result<()> {
+        if !self.in_initial_copy {
+            return Ok(());
+        }
+        // First: commit the copied rows that were added directly to main mem_slice
+        // Use LSN 0 to ensure copied data comes before any CDC events
+        self.commit(0);
+        self.flush(0).await?;
+
+        let commits = self
+            .initial_copy_buffered_commits
+            .drain(..)
+            .collect::<Vec<_>>();
+
+        self.in_initial_copy = false;
+        // Second: apply all buffered streaming transaction commits
+        for commit in commits {
+            let _ = self.commit_transaction_stream(commit).await;
```

Comment:
we should check returned result?

Reply:
If you commit multiple times, like this, snapshot flush LSN will be reset for a few times, which is also a bug.